src/pudl/analysis/timeseries_cleaning.py

```diff
@@ -1560,6 +1561,162 @@ def filter_missing_values(
     return df
 
 
+@dataclass
+class SimulateFlagsSettings:
+    """Define settings used to simulate flagged values for scoring imputation."""
+
+    num_months: int = 30
+    """The number of months of data to simulate."""
+    min_flag_rate: float = 0.1
+    """Min ratio of bad points in a section of data to be used for reference."""
+    max_flag_rate: float = 0.5
+    """Max ratio of bad points in a section of data to be used for reference."""
+
+
+class SimulationDataFrame(pa.DataFrameModel):
+    """Collect months of data which will be used to simulate flagged values.
+
+    Each row in this dataframe will contain a reference ID and month, and a simulation
+    ID and month. The reference ID/month points to a month of reported data where
+    a high ratio of values were flagged for imputation, and the simulation ID/month
+    points to a month of reported data with no flagged values. The flagged values
+    in the reference month will be used to simulate flagged values in the simulation
+    month.
+    """
+
+    reference_id_col: Series[Any]
+    reference_month: Series[pa.dtypes.DateTime]
+    simulation_id_col: Series[Any]
+    simulation_month: Series[pa.dtypes.DateTime]
+
+
+@pa.check_types
+def _get_simulation_data(
+    df: DataFrame[AlignedTimeseriesDataFrame],
+    simulation_df: DataFrame[SimulationDataFrame],
+):
+    """Transform flagged pattern from reference periods to simulation periods.
+
+    This will find all flagged values from a reference month and apply the flag
+    pattern to a simulation month. The flag pattern is determined by calculating the
+    hour of the month for each flagged (how many hours is this after the start of the
+    month), and flagging the corresponding hour in the simulation month. Reference
+    months are chosen by finding months with a relatively high rate of imputation,
+    while simulation months have no values which were flagged for imputation.
+    """
+    # Merge simulation dataframe with aligned timeseries dataframe to get all
+    # Hours in the reference month
+    simulation_df["period"] = simulation_df["reference_month"].dt.to_period("M")
+    df["period"] = df["datetime"].dt.to_period("M")
+
+    simulation_df = df.merge(
+        simulation_df,
+        left_on=["id_col", "period"],
+        right_on=["reference_id_col", "period"],
+        how="inner",
+    )
+
+    # Filter to hours where values where flagged (NULL) in reference month
+    flagged = simulation_df[simulation_df["value_col"].isnull()]
+
+    # Transform hours in reference month to hours in simulation month
+    flagged["simulation_datetime"] = (
+        flagged["datetime"] - flagged["reference_month"] + flagged["simulation_month"]
+    )
```

Comment:
It's not intuitive to me what it means to add and subtract datetimes like this.

Both `simulation_month` and `reference_month` are month-starting datetimes, right?

It seems like `simulation_month` - `reference_month` is meant to be a timedelta -- a shift (an amount of time, rather than a point in time) that can be applied to some other datetime. Is that right?

Why isn't it ever a problem that the time delta is always based on month-start datetimes, when months have different numbers of hours in them, and the time delta is being used to modify hourly timestamps?